# Limpeza dos dados
## Roubo de Celulares

Com os dados obtidos  e já compilados, o próximo passo é realizar a limpeza do dataset
Isso consiste em:
 - Separar a cidade de São Bernardo do Campo
 - Verificar valores nulos
 - Campos duplicados
 - Remover informações que não serão úteis
 - Acertar ou remover informações erradas

### Importar as bibliotecas

As bibliotecas que serão utilizadas são:
 - `Pandas` - Manipulação do dataset
 - `Matplotlib` - Visualização dos dados
 - `Seaborn`- Visualização dos dados
 - `Numpy` - habilita propriedades matemáticas

In [1]:
#importando as bibliotecas
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import os


### Carregar o dataset de roubo de celulares

Carregar o dataset, verificar suas dimensões e ter as primeiras impressões

In [2]:
path = 'data_cleaned/'

directory = os.path.join(path) # Diretório onde estão salvos os arquivos .xls
i=0 # Marcador criado para definir se é o primeiro arquivo ou não
counter=0 #Contador de arquivos lidos


for root,dirs,files in os.walk(directory): # root - Caminho lido | dirs - diretórios existentes dentro do caminho lido | files - arquivos existentes no caminho lido
    
    #Percorrer todos os arquivos um a um, e ler somente os arquivos com extensão .xls
    for file in files:
        if file.endswith("2020.csv"):
            dftemp = pd.read_csv(path+file, low_memory=False) # leitura do arquivo .xls ("corrompido")
            
            if i==0: # Condição IF para determinar se é a primeira leitura e copiar o dataset para a variável "df"
                df = dftemp.copy()
                i = 1
                del(dftemp) # Deletar a variável para não sobrecarregar a memória
                
            else: # Para os demais arquivos realizar a junção com o primeiro arquivo
                df = df.append(dftemp, ignore_index=True)
                del(dftemp) # Deletar a variável para não sobrecarregar a memória
                
            linhas = df.shape[0]
            counter = counter + 1
            
print("Linhas totais: ", linhas)
print('\nTotal de arquivos concatenados:', counter)
print('\nDimensões do dataset: \n\t- {} Linhas\n\t- {} Colunas'.format(df.shape[0],df.shape[1]))

Linhas totais:  6258784

Total de arquivos concatenados: 3

Dimensões do dataset: 
	- 6258784 Linhas
	- 57 Colunas


In [3]:
print('{} Linhas\n{} Colunas'.format(df.shape[0],df.shape[1]))
df.head() # Mostra as 5 primeiras linhas do dataset

6258784 Linhas
57 Colunas


ANO_BO  NUM_BO NUMERO_BOLETIM          BO_INICIADO           BO_EMITIDO  \
0    2009    4329      4329/2009  31/12/2009 23:58:08  01/01/2010 00:01:48   
1    2009    6595      6595/2009  31/12/2009 23:55:54  01/01/2010 00:02:38   
2    2010       1         1/2010  01/01/2010 00:05:13  01/01/2010 00:12:24   
3    2010       1         1/2010  01/01/2010 00:48:15  01/01/2010 00:51:39   
4    2010       1         1/2010  01/01/2010 00:58:51  01/01/2010 01:02:29   

  DATAOCORRENCIA HORAOCORRENCIA PERIDOOCORRENCIA DATACOMUNICACAO  \
0     31/12/2009          23:00          A NOITE      31/12/2009   
1     31/12/2009            NaN          A TARDE      31/12/2009   
2     31/12/2009          19:40          A NOITE      31/12/2009   
3     31/12/2009          23:50          A NOITE      01/01/2010   
4     31/12/2009          22:30          A NOITE      01/01/2010   

        DATAELABORACAO  ... DESCR_COR_VEICULO    DESCR_MARCA_VEICULO  \
0  31/12/2009 23:58:08  ...            Branco         VW/GOL PLUS MI   
1  31/12/2009 23:55:54  ...            Branco               VW/KOMBI   
2  01/01/2010 00:05:13  ...              Azul       FIAT/PALIO YOUNG   
3  01/01/2010 00:48:15  ...             Preta       HONDA/CG 125 FAN   
4  01/01/2010 00:58:51  ...            Branco  IMP/FIAT UNO MILLE SX   

  ANO_FABRICACAO ANO_MODELO DESCR_TIPO_VEICULO QUANT_CELULAR MARCA_CELULAR  \
0         1997.0     1997.0          AUTOMOVEL           NaN           NaN   
1         1974.0     1974.0          AUTOMOVEL           NaN           NaN   
2         2000.0     2001.0          AUTOMOVEL           NaN           NaN   
3         2006.0     2006.0          MOTOCICLO           NaN           NaN   
4         1997.0     1998.0          AUTOMOVEL           NaN           NaN   

  ROUBO_C ROUBO_V FURTO_V  
0       0       0       1  
1       0       0       1  
2       0       0       1  
3       0       0       1  
4       0       0       1  

[5 rows x 57 columns]

### Verificar dados nulos e ausentes

É possivel ver que pelo menos 18 colunas possuem mais de 80% de dados ausentes, a grande maioria relacionado a dados pessoais.

Essas as quais podem ser diretamente removidas do dataset.

Abaixo a lista ordenada em % dos primeiros 35 mais ausentes

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6258784 entries, 0 to 6258783
Data columns (total 57 columns):
 #   Column                    Dtype  
---  ------                    -----  
 0   ANO_BO                    int64  
 1   NUM_BO                    int64  
 2   NUMERO_BOLETIM            object 
 3   BO_INICIADO               object 
 4   BO_EMITIDO                object 
 5   DATAOCORRENCIA            object 
 6   HORAOCORRENCIA            object 
 7   PERIDOOCORRENCIA          object 
 8   DATACOMUNICACAO           object 
 9   DATAELABORACAO            object 
 10  BO_AUTORIA                object 
 11  FLAGRANTE                 object 
 12  NUMERO_BOLETIM_PRINCIPAL  object 
 13  LOGRADOURO                object 
 14  NUMERO                    object 
 15  BAIRRO                    object 
 16  CIDADE                    object 
 17  UF                        object 
 18  LATITUDE                  object 
 19  LONGITUDE                 object 
 20  DESCRICAOLOCAL          

In [5]:
nulos = df.isnull().sum()*100
nulos = nulos / len(df)
nulos.sort_values(ascending=False).head(35)

PARENTESCO                  100.000000
RELACIONAMENTO               99.999105
PROFISSAO                    99.898383
GRAUINSTRUCAO                99.894484
NACIONALIDADE                99.857161
ESTADOCIVIL                  99.847526
NATURALIDADE                 99.838131
DATANASCIMENTO               99.835479
IDADE                        99.830334
SEXO                         99.790183
VITIMAFATAL                  99.786236
TIPOPESSOA                   99.786236
NATUREZAVINCULADA            99.786220
CORCUTIS                     99.786220
TIPOVINCULO                  99.786204
DESDOBRAMENTO                94.816932
NUMERO_BOLETIM_PRINCIPAL     93.829744
EXAME                        83.911731
QUANT_CELULAR                65.123193
MARCA_CELULAR                57.909604
DESCR_MARCA_VEICULO          49.574726
PLACA_VEICULO                49.427684
CIDADE_VEICULO               49.354108
DESCR_TIPO_VEICULO           49.289766
UF_VEICULO                   49.211668
DESCR_COR_VEICULO        

### Remover colunas e dados indesejados

Primeiramente avaliar entre as colunas que tem mais que 70% dos dados ausentes o que é possível eliminar e criar uma lista com esses campos

In [6]:
nulos.index[nulos > 0.7]

Index(['HORAOCORRENCIA', 'NUMERO_BOLETIM_PRINCIPAL', 'LOGRADOURO', 'BAIRRO',
       'LATITUDE', 'LONGITUDE', 'EXAME', 'DESDOBRAMENTO', 'STATUS',
       'TIPOPESSOA', 'VITIMAFATAL', 'NATURALIDADE', 'NACIONALIDADE', 'SEXO',
       'DATANASCIMENTO', 'IDADE', 'ESTADOCIVIL', 'PROFISSAO', 'GRAUINSTRUCAO',
       'CORCUTIS', 'NATUREZAVINCULADA', 'TIPOVINCULO', 'RELACIONAMENTO',
       'PARENTESCO', 'PLACA_VEICULO', 'UF_VEICULO', 'CIDADE_VEICULO',
       'DESCR_COR_VEICULO', 'DESCR_MARCA_VEICULO', 'ANO_FABRICACAO',
       'ANO_MODELO', 'DESCR_TIPO_VEICULO', 'QUANT_CELULAR', 'MARCA_CELULAR'],
      dtype='object')

Os campos pessoais podem ser diretamente eliminados, para os demais campos será analisado a real necessidade

In [7]:
eliminar = ['PARENTESCO', 'RELACIONAMENTO', 'NACIONALIDADE', 'PROFISSAO',
       'GRAUINSTRUCAO', 'NATURALIDADE', 'DATANASCIMENTO', 'ESTADOCIVIL',
       'IDADE', 'NATUREZAVINCULADA', 'TIPOPESSOA', 'SEXO', 'CORCUTIS',
       'TIPOVINCULO']

In [8]:
df.drop(eliminar, axis=1, inplace=True)

In [9]:
len(df.columns)

43

### Analisando demais colunas

Usar comando abaixo para ajustar configurações de visualização do pandas para explorar os dados das colunas

In [10]:
pd.set_option('display.max_columns', 60) # Define limite de 60 colunas na visualização
#pd.set_option('display.max_seq_items', 30) # Define limite de linhas na visualização

Selecionar somente os campos onde há vitima fatal e contar.

De acordo com o site da SSP-SP, a referência de quantidade de Boletins de Ocorrência deve ser pelo Número do Boletim (NUMERO_BOLETIM).


In [11]:
vitimafatal = df[df['VITIMAFATAL'].isnull()==False]
total_vf = len(vitimafatal['NUMERO_BOLETIM'].unique())
print(total_vf, 'vitimas fatais de 2010 à 2020')

2100 vitimas fatais de 2010 à 2020


Neste caso, houve 2100 vitimas fatais entre 2010 e 2020,

Por esta razão os dados serão mantidos, apenas a coluna `'VITIMAFATAL'` não será removida.

O campo `'EXAME'`, `'RUBRICA'`, `'ESPECIE'`, `'DESDOBRAMENTO'`, possuem informações que não são relevantes para a ánalise, porém os demais dados devem ser mantidos, removendos apenas as colunas.


In [12]:
print("EXAME")
print(df.EXAME.unique()[:10],'\n')
print("RUBRICA")
print(df.RUBRICA.unique()[:10],'\n')
print("ESPECIE")
print(df.ESPECIE.unique()[:10],'\n')
print("DESDOBRAMENTO")
print(df.DESDOBRAMENTO.unique()[:10])


EXAME
[nan 'IC' 'IC-IML' 'IML' 'SVO'] 

RUBRICA
['Furto (art. 155) - VEICULO'
 'Furto qualificado (art. 155, §4o.) - RESIDENCIA'
 'Furto qualificado (art. 155, §4o.) - VEICULO'
 'Fuga de local de acidente (Art. 305)'
 'Localização/Apreensão e Entrega de veículo' 'Choque'
 'Entrega de veículo localizado/apreendido'
 'Localização/Apreensão de veículo' 'Ato Infracional'
 'Localização/Apreensão de objeto'] 

ESPECIE
['Título II - Patrimônio (arts. 155 a 183)'
 'L 9503/97 - Código de Trânsito Brasileiro' 'Localização e/ou Devolução'
 'Acidente de trânsito' 'Ato infracional' 'Captura procurado' 'Pessoa'
 'Título XI - Administração pública (arts. 312 a 359-H)'
 'Título VIII - Incolumidade pública (arts. 250 a 285)'
 'Outros - não criminal'] 

DESDOBRAMENTO
[nan '§4o. Se o crime é cometido:'
 'Possuir, deter, portar, adquirir, fornecer, receber, ter em depósito...'
 'caput. Subtrair coisa móvel alheia, mediante grave ameaça ou violência a pessoa'
 'Nas mesmas penas incorrem quem (Par. único)'


In [13]:
eliminar = ["EXAME", "RUBRICA", "ESPECIE", "DESDOBRAMENTO"]

df.drop(eliminar, axis=1, inplace=True)

In [14]:
df.shape

(6258784, 39)

## Analise de variáveis

In [15]:
df.columns

Index(['ANO_BO', 'NUM_BO', 'NUMERO_BOLETIM', 'BO_INICIADO', 'BO_EMITIDO',
       'DATAOCORRENCIA', 'HORAOCORRENCIA', 'PERIDOOCORRENCIA',
       'DATACOMUNICACAO', 'DATAELABORACAO', 'BO_AUTORIA', 'FLAGRANTE',
       'NUMERO_BOLETIM_PRINCIPAL', 'LOGRADOURO', 'NUMERO', 'BAIRRO', 'CIDADE',
       'UF', 'LATITUDE', 'LONGITUDE', 'DESCRICAOLOCAL', 'SOLUCAO',
       'DELEGACIA_NOME', 'DELEGACIA_CIRCUNSCRICAO', 'STATUS', 'VITIMAFATAL',
       'PLACA_VEICULO', 'UF_VEICULO', 'CIDADE_VEICULO', 'DESCR_COR_VEICULO',
       'DESCR_MARCA_VEICULO', 'ANO_FABRICACAO', 'ANO_MODELO',
       'DESCR_TIPO_VEICULO', 'QUANT_CELULAR', 'MARCA_CELULAR', 'ROUBO_C',
       'ROUBO_V', 'FURTO_V'],
      dtype='object')

In [16]:
df.tail()

ANO_BO   NUM_BO NUMERO_BOLETIM          BO_INICIADO  \
6258779    2020  1441970   1441970/2020  30/09/2020 23:09:23   
6258780    2020  1441970   1441970/2020  30/09/2020 23:09:23   
6258781    2020  1440526   1440526/2020  30/09/2020 17:46:46   
6258782    2020  1439915   1439915/2020  30/09/2020 16:25:49   
6258783    2020     2624      2624/2020  30/09/2020 23:19:28   

                  BO_EMITIDO DATAOCORRENCIA HORAOCORRENCIA PERIDOOCORRENCIA  \
6258779  30/09/2020 23:07:52     30/09/2020          20:42          A NOITE   
6258780  30/09/2020 23:07:52     30/09/2020          20:42          A NOITE   
6258781  30/09/2020 23:08:45     30/09/2020          10:45       PELA MANHÃ   
6258782  30/09/2020 23:33:10     29/09/2020          11:20       PELA MANHÃ   
6258783  30/09/2020 23:37:33     30/09/2020          22:30          A NOITE   

        DATACOMUNICACAO       DATAELABORACAO    BO_AUTORIA FLAGRANTE  \
6258779      30/09/2020  30/09/2020 23:09:23  Desconhecida       Não   
6258780      30/09/2020  30/09/2020 23:09:23  Desconhecida       Não   
6258781      30/09/2020  30/09/2020 17:46:46  Desconhecida       Não   
6258782      30/09/2020  30/09/2020 16:25:49  Desconhecida       Não   
6258783      30/09/2020  30/09/2020 23:19:28  Desconhecida       Não   

        NUMERO_BOLETIM_PRINCIPAL                             LOGRADOURO  \
6258779                      NaN            RUA ANDRÉIA DA SILVA BARROS   
6258780                      NaN            RUA ANDRÉIA DA SILVA BARROS   
6258781                      NaN  AVENIDA RAIMUNDO PEREIRA DE MAGALHÃES   
6258782                      NaN           ESTRADA TURÍSTICA DO JARAGUÁ   
6258783                      NaN                         RODOVIA SP 332   

          NUMERO                          BAIRRO           CIDADE  UF  \
6258779    347.0  JARDIM TERRAS DE SANTO ANTONIO      HORTOLANDIA  SP   
6258780    347.0  JARDIM TERRAS DE SANTO ANTONIO      HORTOLANDIA  SP   
6258781  12495.0                   JARDIM MARILU          S.PAULO  SP   
6258782    700.0                    VILA JARAGUÁ          S.PAULO  SP   
6258783     45.0                  CRISTAL PARQUE  FRANCO DA ROCHA  SP   

                  LATITUDE          LONGITUDE  \
6258779  -22,9062783537479  -47,1787835739918   
6258780  -22,9062783537479  -47,1787835739918   
6258781        -23,4307817        -46,7203606   
6258782        -23,4880678        -46,7554815   
6258783        -23,2960414      -46,777100127   

                          DESCRICAOLOCAL                         SOLUCAO  \
6258779  Garagem ou abrigo de residência            BO PARA INVESTIGAÇÃO   
6258780  Garagem ou abrigo de residência            BO PARA INVESTIGAÇÃO   
6258781                      Via pública                BO PARA REGISTRO   
6258782                      Via pública  ENCAMINHAMENTO DP ÁREA DO FATO   
6258783                      Via pública  ENCAMINHAMENTO DP ÁREA DO FATO   

                   DELEGACIA_NOME DELEGACIA_CIRCUNSCRICAO     STATUS  \
6258779      DELEGACIA ELETRONICA    01º D.P. HORTOLÂNDIA  Consumado   
6258780      DELEGACIA ELETRONICA    01º D.P. HORTOLÂNDIA  Consumado   
6258781      DELEGACIA ELETRONICA          46º D.P. PERUS  Consumado   
6258782      DELEGACIA ELETRONICA          46º D.P. PERUS  Consumado   
6258783  DEL.POL.FRANCISCO MORATO    DEL.POL.FRANCO ROCHA  Consumado   

        VITIMAFATAL PLACA_VEICULO UF_VEICULO CIDADE_VEICULO DESCR_COR_VEICULO  \
6258779         NaN           NaN         SP            NaN             Preta   
6258780         NaN       DPL1580         SP       CAMPINAS          Vermelho   
6258781         NaN       EVI5936         SP       VALINHOS            Branco   
6258782         NaN       FSK6835         SP      SÃO PAULO            Branco   
6258783         NaN       EFL8943         SP        CAJAMAR          Vermelho   

            DESCR_MARCA_VEICULO  ANO_FABRICACAO  ANO_MODELO  \
6258779      HONDA/CG 125 TITAN             NaN         NaN   
6258780       YAMAHA/FZ25 FA

In [17]:
df.dtypes

ANO_BO                        int64
NUM_BO                        int64
NUMERO_BOLETIM               object
BO_INICIADO                  object
BO_EMITIDO                   object
DATAOCORRENCIA               object
HORAOCORRENCIA               object
PERIDOOCORRENCIA             object
DATACOMUNICACAO              object
DATAELABORACAO               object
BO_AUTORIA                   object
FLAGRANTE                    object
NUMERO_BOLETIM_PRINCIPAL     object
LOGRADOURO                   object
NUMERO                       object
BAIRRO                       object
CIDADE                       object
UF                           object
LATITUDE                     object
LONGITUDE                    object
DESCRICAOLOCAL               object
SOLUCAO                      object
DELEGACIA_NOME               object
DELEGACIA_CIRCUNSCRICAO      object
STATUS                       object
VITIMAFATAL                  object
PLACA_VEICULO                object
UF_VEICULO                  

### Colunas que devem ser convertidas para datas / inteiros / categoricos

In [18]:
datas = ['BO_INICIADO', 'BO_EMITIDO',
       'DATAOCORRENCIA', 'HORAOCORRENCIA',
       'DATACOMUNICACAO', 'DATAELABORACAO']
inteiros = ['NUMERO', 'ANO_FABRICACAO', 'ANO_MODELO', 'QUANT_CELULAR']

categoricos = ['PERIDOOCORRENCIA', 'BO_AUTORIA', 'FLAGRANTE',
       'DESCRICAOLOCAL', 'SOLUCAO', 'STATUS',
       'VITIMAFATAL', 'DESCR_COR_VEICULO',
       'DESCR_TIPO_VEICULO']


#### Variáveis Categóricas

In [19]:
df[categoricos] = df[categoricos].astype("category")

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6258784 entries, 0 to 6258783
Data columns (total 39 columns):
 #   Column                    Dtype   
---  ------                    -----   
 0   ANO_BO                    int64   
 1   NUM_BO                    int64   
 2   NUMERO_BOLETIM            object  
 3   BO_INICIADO               object  
 4   BO_EMITIDO                object  
 5   DATAOCORRENCIA            object  
 6   HORAOCORRENCIA            object  
 7   PERIDOOCORRENCIA          category
 8   DATACOMUNICACAO           object  
 9   DATAELABORACAO            object  
 10  BO_AUTORIA                category
 11  FLAGRANTE                 category
 12  NUMERO_BOLETIM_PRINCIPAL  object  
 13  LOGRADOURO                object  
 14  NUMERO                    object  
 15  BAIRRO                    object  
 16  CIDADE                    object  
 17  UF                        object  
 18  LATITUDE                  object  
 19  LONGITUDE                 object  
 20  DE

#### Variáveis Númericas

In [21]:
df.select_dtypes(include=['integer', 'float']).columns

Index(['ANO_BO', 'NUM_BO', 'ANO_FABRICACAO', 'ANO_MODELO', 'ROUBO_C',
       'ROUBO_V', 'FURTO_V'],
      dtype='object')

Verificar se os valores de cada uma das colunas estão consistentes

##### ANO_BO

In [23]:
df['ANO_BO'].unique()

array([2009, 2010, 2011, 2012, 2017, 2013, 2016, 2014, 2015, 2018, 2019,
       2020], dtype=int64)

##### NUM_BO

In [24]:
df.loc[df['NUM_BO'] < 0]

Empty DataFrame
Columns: [ANO_BO, NUM_BO, NUMERO_BOLETIM, BO_INICIADO, BO_EMITIDO, DATAOCORRENCIA, HORAOCORRENCIA, PERIDOOCORRENCIA, DATACOMUNICACAO, DATAELABORACAO, BO_AUTORIA, FLAGRANTE, NUMERO_BOLETIM_PRINCIPAL, LOGRADOURO, NUMERO, BAIRRO, CIDADE, UF, LATITUDE, LONGITUDE, DESCRICAOLOCAL, SOLUCAO, DELEGACIA_NOME, DELEGACIA_CIRCUNSCRICAO, STATUS, VITIMAFATAL, PLACA_VEICULO, UF_VEICULO, CIDADE_VEICULO, DESCR_COR_VEICULO, DESCR_MARCA_VEICULO, ANO_FABRICACAO, ANO_MODELO, DESCR_TIPO_VEICULO, QUANT_CELULAR, MARCA_CELULAR, ROUBO_C, ROUBO_V, FURTO_V]
Index: []

##### NUMERO

In [25]:
df = df[(df['NUMERO'] != 'Título II - Patrimônio (arts. 155 a 183)') & (df['NUMERO'] != '96º D.P. MONÇÕES')]

In [26]:
for i in inteiros:
    print(i)
    df[i] = pd.to_numeric(df[i].replace(np.nan, 0), downcast='integer', errors='coerce')

NUMERO
ANO_FABRICACAO
ANO_MODELO
QUANT_CELULAR


Na coluna número é possivel ver que tem alguns números negativos, para isso vamos considerá-los como positivos

In [27]:
pd.set_option('display.float_format', lambda x: '%.0f' % x)

In [28]:
df['NUMERO'].describe()

count      6258782
mean         27959
std        5103881
min         -19739
25%              0
50%            100
75%            440
max     1998810785
Name: NUMERO, dtype: float64

In [29]:
len(df.loc[df['NUMERO'] < 0])

7

In [30]:
df['NUMERO'] = np.where((df['NUMERO'] < 0), abs(df['NUMERO']), df['NUMERO'])

In [31]:
df.loc[df['NUMERO'] < 0]

Empty DataFrame
Columns: [ANO_BO, NUM_BO, NUMERO_BOLETIM, BO_INICIADO, BO_EMITIDO, DATAOCORRENCIA, HORAOCORRENCIA, PERIDOOCORRENCIA, DATACOMUNICACAO, DATAELABORACAO, BO_AUTORIA, FLAGRANTE, NUMERO_BOLETIM_PRINCIPAL, LOGRADOURO, NUMERO, BAIRRO, CIDADE, UF, LATITUDE, LONGITUDE, DESCRICAOLOCAL, SOLUCAO, DELEGACIA_NOME, DELEGACIA_CIRCUNSCRICAO, STATUS, VITIMAFATAL, PLACA_VEICULO, UF_VEICULO, CIDADE_VEICULO, DESCR_COR_VEICULO, DESCR_MARCA_VEICULO, ANO_FABRICACAO, ANO_MODELO, DESCR_TIPO_VEICULO, QUANT_CELULAR, MARCA_CELULAR, ROUBO_C, ROUBO_V, FURTO_V]
Index: []

Também é possivel identificar diversos números muito maiores do que o encontrados normalmente em nossas ruas, mas fica impossivel saber qual seria a numeração correta

primeiramente será feito um filtro onde todos os números maiors do que 9 mil serão avaliados e contados

In [32]:
out_num = df['NUMERO'].loc[df['NUMERO'] > 9000].value_counts().sum()
out_num

77718

In [33]:
#20 primeiros registros com números maiores do que 9mil
df['NUMERO'].loc[df['NUMERO'] > 9000].value_counts().head(20)

99999    15463
9999      6957
10000     3818
12000     2192
15000     1743
11000     1496
11111     1252
13000     1004
22540      956
16000      932
14000      819
18000      668
20000      665
25000      653
17000      653
16741      645
10500      606
11500      555
9500       531
22000      469
Name: NUMERO, dtype: int64

In [34]:
out_num_p100 = out_num*100 / df.shape[0] 
print('{}% dos dados possuem número da rua maior do que 9 mil'.format(round(out_num_p100, 2)))

1.24% dos dados possuem número da rua maior do que 9 mil


Por ser uma quantia não tão significante, esses números considerados como número "1"


In [35]:
df['NUMERO'] = np.where((df['NUMERO'] > 9000), 1, df['NUMERO'])

Ao mesmo tempo existem mais de 1,73 milhão de registros com o número zero (27,64%), estes também serão passados para "1"

In [36]:
print(df['NUMERO'].loc[df['NUMERO'] == 0].value_counts().sum(), 'entradas com valor "zero"')
print(round(100*df['NUMERO'].loc[df['NUMERO'] == 0].value_counts().sum() / df.shape[0],2), '%')


1730148 entradas com valor "zero"
27.64 %


In [37]:
df['NUMERO'] = np.where((df['NUMERO'] == 0 ), 1, df['NUMERO'])

In [38]:
df['NUMERO'].describe()

count   6258782
mean        450
std         998
min           1
25%           1
50%         100
75%         402
max        9000
Name: NUMERO, dtype: float64

#####  ANO_FABRICACAO

Podemos verificar que há diversos anos de fabricação que não consistem com a realidade, uma vez que o primeiro automovel foi fabricado em 1886

Alguns deste valores como apenas "8" poderia representar "2008", outras como "95" poderiam ser "1995", mas como não tem como ter certeza, não podemos assumir isso logo de primeira.

O primeiro carro fabricado no Brasil foi no ano de 1956, pela Romi-sIetta


In [39]:
df['ANO_FABRICACAO'].unique()

array([1997, 1974, 2000, 2006,    0, 2008, 1994, 2007, 1989, 1985, 2002,
       2005, 1990, 1992, 1984, 1995, 1991, 1999, 2001, 2003, 1987, 2004,
       1996, 1988, 1983, 1993, 1982, 1998, 1975, 2009, 1979, 1976, 1986,
       1977, 1980, 1970, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972, 1967, 1965, 2010,  200,  198, 1098, 2101, 1962, 2011, 1961,
         84, 1081, 1963, 1957,    7,    4, 1952, 1959,   89,  209, 2099,
       1960, 1841,   91, 2088, 1958,   98,   10,  978, 1951, 1899,  201,
       1928, 2012,  985, 2066, 2013,   12, 2998,   95, 8989, 2207,   97,
       2911, 2021, 2014, 3013, 2912, 1190,  202, 9394,   87,  992, 1085,
       1948,  211,  708, 1112, 1013, 2104, 2103, 1898, 2015, 2913, 1093,
       1191, 1195,  193, 1910,  213, 1193, 1079, 1883, 1083, 1944, 1889,
        910, 1111,  997, 1515, 1414,  304, 1194,  607, 1415, 1011, 1940,
       1199, 2206, 1010,   73, 1945, 1213, 1088,   99, 1914,    6,   96,
       1947, 7998,  809,   93, 1884, 1903, 2016,  2


- Primeiro passo será separar todas as entradas menor que 1950 e maior que 2020
- Depois contar qual a quantidade e a porcentagem destas entradas em relação aos demais
- Analisar em detalhes alguns destes valores por amostragem

In [40]:
#seleciona os campos onde Ano de Fabricação é menor que 1950 e maior que 2020
ano_fab = df.query('ANO_FABRICACAO < 1950 or ANO_FABRICACAO > 2020')

#Seleciona os campos onde Ano de Fabricação é diferente de zero
ano_fab = ano_fab.loc[df.ANO_FABRICACAO != 0]
print(len(ano_fab), 'Entradas onde Ano de fabricação é menor que 1950, maior que 2020 e diferente de zero')
print('{}% do total de entradas'.format(round(len(ano_fab)*100 / df.shape[0],2)))

467 Entradas onde Ano de fabricação é menor que 1950, maior que 2020 e diferente de zero
0.01% do total de entradas


In [41]:
# seleciona 20 amostras aleatórias para uma avaliação mais detalhada
ano_fab[['ANO_BO', 'PLACA_VEICULO', 'DESCR_MARCA_VEICULO', 'DESCR_TIPO_VEICULO', 'ANO_FABRICACAO', 'ANO_MODELO']].sample(n=20, random_state=111)

ANO_BO PLACA_VEICULO      DESCR_MARCA_VEICULO DESCR_TIPO_VEICULO  \
265403     2012       CFJ8507           H/HONDA CG 125          MOTOCICLO   
5881117    2017       PZN3268          FIAT/FIAT UNO S          AUTOMOVEL   
6232869    2020       LRF8342   I/BMW X5 M COMPETITION          AUTOMOVEL   
4655453    2011       MFE2419   SR/RODOTECNICA SRT TQ2       SEMI-REBOQUE   
1921541    2011       AMJ8941           M.BENZ/LS 1634    CAMINHÃO TRATOR   
1456835    2020       DYZ2042              H/HONDA 150          MOTOCICLO   
1994237    2012       AQL1447      VW/25.370 CLM T 6X2    CAMINHÃO TRATOR   
2946306    2015       FMQ1566                FIAT/FIAT          AUTOMOVEL   
2356315    2013       AWI7907       FIAT/MILLE ECONOMY          AUTOMOVEL   
2356317    2013       AWI7907       FIAT/MILLE ECONOMY          AUTOMOVEL   
5506106    2015       FMQ1566                FIAT/FIAT          AUTOMOVEL   
196073     2011       HLG1524  FIAT/PALIO FIRE ECONOMY          AUTOMOVEL   
4899998    2012       NZO3112                IMP/VOLVO           CAMINHÃO   
4640950    2011       GMQ6569            FORD/ESCORT L          AUTOMOVEL   
1234830    2018       IJM0929                 IMP/FORD          AUTOMOVEL   
123214     2010           HBT    HONDA/CBX 250 TWISTER          MOTOCICLO   
1971234    2012       NKA5421    REB/NOMA SR2E18RT1 CG       SEMI-REBOQUE   
4498844    2010       KJS7088     HONDA/XR 250 TORNADO          MOTOCICLO   
1009939    2016       FJT5760  CHEVROLET/CELTA 1.0L LT          AUTOMOVEL   
1296448    2018       BGS9643        IMP/VW GOL GL 1.8          AUTOMOVEL   

         ANO_FABRICACAO  ANO_MODELO  
265403              985        1985  
5881117            2107        2018  
6232869              13           0  
4655453             207        2007  
1921541            5004        2004  
1456835             200           0  
1994237             208        2008  
2946306             213         214  
2356315            7907        2012  
2356317            7907        2012  
5506106             213         214  
196073               10           0  
4899998            1011           0  
4640950              84          84  
1234830            1930        1930  
123214             1841           0  
1971234            2207        2008  
4498844             205           0  
1009939            1314        1314  
1296448              92          92

Conforme observado na tabela acima se analisado o campo `DESCR_MARCA_VEICULO`, `ANO_FABRICACAO` e `ANO_MODELO`, a maioria dos casos está claro que foram erros de digitação e poderiam ser facilmente arrumados.

No entanto por se tratar apenas de 0,01% de todos os registros esses valores serão considerados como "0"

In [42]:
#seleciona a coluna ANO_FABRICACAO onde o index é igual do dataset "ano_fab" e define os valores como "0"
df['ANO_FABRICACAO'].loc[ano_fab.index] = 0

c:\users\caior\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [43]:
df['ANO_FABRICACAO'].unique()

array([1997, 1974, 2000, 2006,    0, 2008, 1994, 2007, 1989, 1985, 2002,
       2005, 1990, 1992, 1984, 1995, 1991, 1999, 2001, 2003, 1987, 2004,
       1996, 1988, 1983, 1993, 1982, 1998, 1975, 2009, 1979, 1976, 1986,
       1977, 1980, 1970, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972, 1967, 1965, 2010, 1962, 2011, 1961, 1963, 1957, 1952, 1959,
       1960, 1958, 1951, 2012, 2013, 2014, 2015, 2016, 1956, 2017, 1955,
       2018, 1954, 2019, 2020], dtype=int16)

#####  ANO_MODELO

Repetir os mesmos passos do `ANO_FABRICACAO` mas com `ANO_MODELO` até 2021

com o comando `df.ANO_MODELO.unique()` já é possivel identificar que há divergências nos valores 


In [44]:
df.ANO_MODELO.unique()

array([1997, 1974, 2001, 2006, 1998,    0, 2008, 2007, 1994, 1989, 1985,
       2002, 2005, 1990, 1992, 1984, 1996, 1991, 2000, 2009, 2003, 1987,
       1995, 2004, 1988, 1983, 1999, 1986, 1993, 1982, 1975, 1979, 1976,
       1977, 2010, 1980, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972,  125, 1600, 1970, 1967, 1965,  313, 5610, 2011, 1962, 4000,
       4013, 1218,  100, 1961,   84,   40, 1963, 1957,    8, 2020,    4,
       1952,  275, 1620, 1720, 1959, 1155,   89, 8140, 7630,  290, 1684,
       1634,  150, 1960, 2855,  500,    6,  250, 2012,   16, 5516,   91,
       3020,  580,  200,  285,  450, 1016, 1606, 1958,  400,   98, 1000,
       1951,  980, 1928, 9996,   10,    5, 2013, 2408,   12, 1111, 1518,
       2322,   97, 2207, 3514, 8150, 2208, 2066, 2014, 1780, 2913,   93,
       1190, 4283, 9394,  350, 1195,   88,   13, 3013,  815,    1, 9600,
       4600,  416,  292, 1948,  213,  112, 1013, 2015,  708,  750, 1112,
       1898,  265, 2103, 1093,  320, 1191, 2033, 19


- Primeiro passo será separar todas as entradas menor que 1950 e maior que 2021
- Depois contar qual a quantidade e a porcentagem destas entradas em relação aos demais
- Analisar em detalhes alguns destes valores por amostragem

In [45]:
#seleciona os campos onde Ano de Fabricação é menor que 1950 e maior que 2020
ano_mod = df.query('ANO_MODELO < 1950 or ANO_MODELO > 2021')

#Seleciona os campos onde Ano de Fabricação é diferente de zero
ano_mod = ano_mod.loc[df.ANO_MODELO != 0]
print(len(ano_mod), 'Entradas onde o ano do modelo é menor que 1950, maior que 2021 e diferente de zero')
print('{}% do total de entradas'.format(round(len(ano_mod)*100 / df.shape[0],2)))

1177 Entradas onde o ano do modelo é menor que 1950, maior que 2021 e diferente de zero
0.02% do total de entradas


In [46]:
# seleciona 20 amostras aleatórias para uma avaliação mais detalhada
ano_mod[['ANO_BO', 'PLACA_VEICULO', 'DESCR_MARCA_VEICULO', 'DESCR_TIPO_VEICULO', 'ANO_FABRICACAO', 'ANO_MODELO']].sample(n=20, random_state=111)

ANO_BO PLACA_VEICULO      DESCR_MARCA_VEICULO DESCR_TIPO_VEICULO  \
4755486    2011       MIJ9299                IMP/VOLVO           CAMINHÃO   
5480183    2015           NaN       I/SHINERAY XY 50 Q           MOTONETA   
2311407    2013       LLN9739  I/LR DISCOVERY 4 3.0 SE          AUTOMOVEL   
1667813    2010       BXC2856             REB/ANTONINI       SEMI-REBOQUE   
2013088    2012       HJR9607                VW/24.250           CAMINHÃO   
4970816    2013       NYR6395   SCANIA/P340 BAULER CAM    CAMINHÃO TRATOR   
4574541    2010       MFX5294        VOLVO/FH 440 6X2T       TRATOR RODAS   
3618602    2017           NaN          M.A./IMP TRATOR       TRATOR RODAS   
763734     2015           NaN     M.A./MASSEY-FERGUSON       TRATOR RODAS   
5527478    2015           NaN       I/SHINERAY XY 50 Q          CICLOMOTO   
2413752    2014       GSW1004               I/MERCEDES           CAMINHÃO   
551548     2013       GMZ1043         FORD/FORD F 4000           CAMINHÃO   
152542     2011       0000000          M.A./IMP TRATOR       TRATOR RODAS   
142674     2011       HMO8532                 GM/CELTA          AUTOMOVEL   
4793520    2012       GVJ6827  IMP/SCANIA T113H4X2 360    CAMINHÃO TRATOR   
1149768    2017           NaN         MA/MASSEY MF4265       TRATOR MISTO   
5239577    2014       FZX3635  CHEVROLET/ONIX 1.0MT LT          AUTOMOVEL   
326417     2012       COH1426            GM/KADETT GLS          AUTOMOVEL   
2658770    2014       EZL8012      I/KIA KM BETA BONGO           CAMINHÃO   
161548     2011       MSW2248              VW/CROSSFOX          AUTOMOVEL   

         ANO_FABRICACAO  ANO_MODELO  
4755486            2010         440  
5480183            2013        1014  
2311407            2011         211  
1667813            1993        1933  
2013088            2011        2425  
4970816               0         340  
4574541            2008         440  
3618602               0        6615  
763734             2000         275  
5527478            2015          50  
2413752            2011        1620  
551548             1989        4000  
152542             1980         285  
142674             2010        3020  
4793520               0          98  
1149768               0         265  
5239577            2013          10  
326417                0          97  
2658770               0          12  
161548             2009        1606

Como já "corrigimos" toda a coluna "ANO_FABRICACAO", vamos considerar que para todas essas entradas divergente que:
```python
ANO_MODELO = ANO_FABRICACAO
```


In [47]:
ano_mod['ANO_MODELO'] = ano_mod['ANO_FABRICACAO']

In [48]:
#seleciona a coluna ANO_FABRICACAO onde o index é igual do dataset "ano_fab" e define os valores como "0"
df['ANO_MODELO'].loc[ano_mod.index] = ano_mod['ANO_MODELO']

c:\users\caior\appdata\local\programs\python\python39\lib\site-packages\pandas\core\indexing.py:1637: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [49]:
df.ANO_MODELO.unique()

array([1997, 1974, 2001, 2006, 1998,    0, 2008, 2007, 1994, 1989, 1985,
       2002, 2005, 1990, 1992, 1984, 1996, 1991, 2000, 2009, 2003, 1987,
       1995, 2004, 1988, 1983, 1999, 1986, 1993, 1982, 1975, 1979, 1976,
       1977, 2010, 1980, 1971, 1978, 1973, 1969, 1981, 1964, 1966, 1968,
       1972, 1970, 1967, 1965, 2011, 1962, 1961, 1963, 1957, 2020, 1952,
       1959, 1960, 2012, 1958, 1951, 2013, 2014, 2015, 2016, 2017, 1956,
       2018, 1955, 2019, 1954, 2021], dtype=int16)

#####  QUANT_CELULAR

Como foi utilizado o comando 

```python
errors = 'coerce'
```
Os valores que estavam como inválidos passaram a ser considerados NaN, é possível verificar que apenas 7 entradas estavam com essa inconsistência, por este motivo as mesmas serão eliminadas



In [50]:
df.QUANT_CELULAR.isnull().sum()

7

In [51]:
df.dropna(subset = ['QUANT_CELULAR'], inplace=True)

Verificar as métricas estatisticas das quantidades de celulares roubados, apesar de aparecer alguns outliers, estes serão mantidos por poderem se tratar de roubo de carga

In [52]:
# resumo estatistico onde a quantidade de celular é diferente de zero
df.QUANT_CELULAR.loc[df.QUANT_CELULAR != 0].describe()

count    2181781
mean           8
std         9972
min            1
25%            1
50%            1
75%            1
max     14728864
Name: QUANT_CELULAR, dtype: float64

Conforme tabela acima, é possivel ver que 75% das ocorrências acontecem com o roubo de apenas 1 celular,

Para verificação de uma amostragem, será considerado apenas quantidades que estão além dos 8 sigmas (99,9994%)

In [53]:
print('99,9994% das quantidades estão abaixo de:', df.QUANT_CELULAR.loc[df.QUANT_CELULAR != 0].quantile(0.999994))

99,9994% das quantidades estão abaixo de: 5760.0


Apenas 12 entradas estão acima de 5760 unidades roubadas, e somente 2 entradas estão com valores discrepantes dos demais, portanto essas 2 entradas serão eliminadas

index:
 - 3411515
 - 3446741

In [54]:
df[['ANO_BO', 'NUM_BO', 'NUMERO_BOLETIM_PRINCIPAL', 'HORAOCORRENCIA', 'CIDADE', 'DESCRICAOLOCAL',
    'QUANT_CELULAR', 'MARCA_CELULAR']]\
.loc[df.QUANT_CELULAR > df.QUANT_CELULAR.loc[df.QUANT_CELULAR != 0].quantile(0.999994)]

ANO_BO  NUM_BO NUMERO_BOLETIM_PRINCIPAL HORAOCORRENCIA  \
2496432    2014   12152                      NaN          16:00   
2508773    2014   14257                      NaN          15:00   
2620589    2014    4646                      NaN          22:20   
2724402    2014    4850                      NaN          00:30   
2898922    2015    1981                      NaN          06:15   
2900938    2015    2031        1981/2015 - 30127          06:15   
3285240    2016    1924                      NaN          18:30   
3411515    2017    1431                      NaN          01:30   
3446741    2017    1984                      NaN          10:30   
3742202    2018    2900                      NaN          16:55   
3743487    2018    1883       2900/2018 - 100908          16:55   
4337546    2020     594                      NaN          12:00   

                        CIDADE       DESCRICAOLOCAL  QUANT_CELULAR  \
2496432                S.PAULO          Via pública           9413   
2508773    S.BERNARDO DO CAMPO          Via pública           9675   
2620589                S.PAULO           Residência           9860   
2724402                DIADEMA          Via pública           9579   
2898922                DIADEMA          Via pública           9844   
2900938                DIADEMA          Via pública           9844   
3285240                S.PAULO  Comércio e serviços           9999   
3411515  FERRAZ DE VASCONCELOS          Via pública         168455   
3446741                 SUMARE          Via pública       14728864   
3742202             ARARAQUARA  Comércio e serviços           9999   
3743487             ARARAQUARA  Comércio e serviços           9999   
4337546            SERTAOZINHO  Comércio e serviços           9999   

              MARCA_CELULAR  
2496432             ALCATEL  
2508773             MOTORLA  
2620589            MOTOROLA  
2724402                 NaN  
2898922  SANSUNG GALAXY DUO  
2900938  SANSUNG GALAXY DUO  
3285240            DIVERSAS  
3411515             Samsung  
3446741                Asus  
3742202                 NaN  
3743487                 NaN  
4337546                 NaN

In [55]:
df.drop([3411515,3446741], axis=0, inplace=True)

## Salvar o arquivo em um  arquivo PARQUET

In [56]:
df.to_parquet('data_cleaned/ssp_cleaned.parquet', index=False)

## Salvar o arquivo em um novo arquivo *.csv*

In [57]:
df.to_csv('data_cleaned/ssp_cleaned.csv', index=False)